### PR TITLE
cds: migrate CDS to Subscription model.

### PR DESCRIPTION
**Pull request by htuch**
Legacy REST adapter translates REST JSON API fetches to a v2 Subscription, similar to SDS/EDS.

### CODE REVIEW

File: source/common/upstream/cds_api_impl.cc [hunk 1]
```diff
@@ -1,65 +1,53 @@
 #include "common/upstream/cds_api_impl.h"
 
-#include <chrono>
 #include <string>
-#include <vector>
 
-#include "common/common/assert.h"
-#include "common/config/cds_json.h"
+#include "common/config/subscription_factory.h"
```
Comment (mattklein123) [inline source/common/upstream/cds_api_impl.cc]: nit: I think BUILD file can be pruned for this?

File: source/common/upstream/cds_api_impl.cc [hunk 2]
```diff
@@ -1,65 +1,53 @@
...
+#include "common/upstream/cds_subscription.h"
 
 namespace Envoy {
 namespace Upstream {
 
 CdsApiPtr CdsApiImpl::create(const Json::Object& config, const Optional<SdsConfig>& sds_config,
                              ClusterManager& cm, Event::Dispatcher& dispatcher,
                              Runtime::RandomGenerator& random,
-                             const LocalInfo::LocalInfo& local_info, Stats::Scope& scope) {
+                             const LocalInfo::LocalInfo& local_info, Stats::Store& store) {
```
Comment (mattklein123) [inline source/common/upstream/cds_api_impl.cc]: Why the switch from scope to store? (Optimally all code just using stats at this point should be passed a scope).

File: source/common/upstream/cds_api_impl.cc [hunk 3]
```diff
@@ -1,65 +1,53 @@
...
-    : RestApiFetcher(cm, config.getObject("cluster")->getString("name"), dispatcher, random,
-                     std::chrono::milliseconds(config.getInteger("refresh_delay_ms", 30000))),
-      local_info_(local_info),
-      stats_({ALL_CDS_STATS(POOL_COUNTER_PREFIX(scope, "cluster_manager.cds."))}),
-      sds_config_(sds_config) {
-  Config::Utility::checkLocalInfo("cds", local_info);
+CdsApiImpl::CdsApiImpl(const envoy::api::v2::ConfigSource& cds_config,
+                       const Optional<SdsConfig>& sds_config, ClusterManager& cm,
+                       Event::Dispatcher& dispatcher, Runtime::RandomGenerator& random,
+                       const LocalInfo::LocalInfo& local_info, Stats::Store& store)
```
Comment (mattklein123) [inline source/common/upstream/cds_api_impl.cc]: store -> scope?

File: source/common/upstream/cds_api_impl.cc [hunk 4]
```diff
@@ -70,24 +58,22 @@ void CdsApiImpl::parseResponse(const Http::Message& response) {
     }
   }
 
-  stats_.update_success_.inc();
+  runInitializeCallbackIfAny();
+}
+
+void CdsApiImpl::onConfigUpdateFailed(const EnvoyException* e) {
+  UNREFERENCED_PARAMETER(e);
```
Comment (mattklein123) [inline source/common/upstream/cds_api_impl.cc]: nit: most code in this case just deletes e in the signature and then you don't need the macro.

File: source/common/upstream/cds_subscription.cc [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+  callbacks_->onConfigUpdate(resources);
+  stats_.update_success_.inc();
+}
+
+void CdsSubscription::onFetchComplete() {}
+
+void CdsSubscription::onFetchFailure(const EnvoyException* e) {
+  callbacks_->onConfigUpdateFailed(e);
+  stats_.update_failure_.inc();
```
Comment (mattklein123) [inline source/common/upstream/cds_subscription.cc]: minor thought: I feel like some of this logic will be the same when the native gRPC impl is done, but I guess we can figure that out when the time comes. (Probably same for EDS).

File: test/common/upstream/cds_api_impl_test.cc [hunk 6]
```diff
@@ -86,9 +85,10 @@ class CdsApiImplTest : public testing::Test {
   Stats::IsolatedStoreImpl store_;
   Http::MockAsyncClientRequest request_;
   CdsApiPtr cds_;
-  Event::MockTimer* interval_timer_{new Event::MockTimer(&dispatcher_)};
```
Comment (mattklein123) [inline test/common/upstream/cds_api_impl_test.cc]: Q: Why is this change necessary?

File: source/common/upstream/cds_api_impl.cc (hunk 1 above)
Comment (htuch) [inline source/common/upstream/cds_api_impl.cc]: I walked through the headers and validated that all the BUILD deps are required. The JSON related ones will go away soon when we drop any need for `Json::Object` in this module.

File: source/common/upstream/cds_api_impl.cc (hunk 2 above)
Comment (htuch) [inline source/common/upstream/cds_api_impl.cc]: I needed to be able to generate a scope with the `cluster_managed.cds` prefix to pass to `Config:Utility::generateStats` below.

File: source/common/upstream/cds_subscription.cc (hunk 5 above)
Comment (htuch) [inline source/common/upstream/cds_subscription.cc]: This is already done in https://github.com/lyft/envoy/blob/master/source/common/config/grpc_subscription_impl.h#L127

File: test/common/upstream/cds_api_impl_test.cc (hunk 6 above)
Comment (htuch) [inline test/common/upstream/cds_api_impl_test.cc]: Creating the `MockTimer` sets up some `EXPECT_CALL` which doesn't apply in the `CdsApiImplTest.InvalidOptions` case. I think we maybe reject bad configurations earlier than this call now.

File: source/common/upstream/cds_api_impl.cc (hunk 1 above)
Comment (mattklein123) [inline source/common/upstream/cds_api_impl.cc]: It doesn't really matter to me but shouldn't most of the includes be transitive? "cds_api_lib" depends on a lot of stuff...

Comment (mattklein123) [inline source/common/upstream/cds_api_impl.cc]: Sigh nevermind me. Sorry. Wasn't looking at header file. Doing too much at once.

File: source/common/upstream/cds_subscription.cc (hunk 5 above)
Comment (mattklein123) [inline source/common/upstream/cds_subscription.cc]: I mostly meant there is some duplication of logic, but nbd.

File: source/common/upstream/cds_api_impl.cc (hunk 1 above)
Comment (htuch) [inline source/common/upstream/cds_api_impl.cc]: Yeah. The rule-of-thumb we're trying to do I think is IWYU (although admittedly, this is rarely complete as we don't have automated enforcement). BUILD deps should largely reflect the headers in the `.{h,cc}`.

File: source/common/upstream/cds_subscription.cc (hunk 5 above)
Comment (htuch) [inline source/common/upstream/cds_subscription.cc]: Yep, it was either a question of doing it once per Subscription or once per xDS API consumer. It seemed to make sense to do it once per Subscription, as there are only 3 of these in v2. However, that ignored the v1 adapters. Since these are (in theory) transient and will go away one day, it seems like something that will improve.
